apps/src/lib/node/ledger/events/log.rs

```diff
@@ -0,0 +1,893 @@
+//! A log to store events emitted by `FinalizeBlock` calls in the ledger.
+//!
+//! The log is flushed every other `N` block heights, or every other `E`
+//! `FinalizeBlock` events, where `N` and `E` are configurable parameters.
+
+pub mod dumb_queries;
+
+use std::default::Default;
+use std::ops::ControlFlow;
+use std::sync::{Arc, RwLock};
+
+use namada::types::storage::BlockHeight;
+use tokio::sync::mpsc::{self, UnboundedReceiver, UnboundedSender};
+use tokio::time::Instant;
+
+use crate::node::ledger::events::Event;
+
+/// Run a CPU-bound task without blocking the Tokio runtime.
+macro_rules! block_in_place {
+    ($expr:expr) => {{
+        // we need this because `tokio_test` panics if we
+        // call `tokio::task::block_in_place()`
+        #[cfg(test)]
+        {
+            $expr
+        }
+        #[cfg(not(test))]
+        {
+            ::tokio::task::block_in_place(|| $expr)
+        }
+    }};
+}
+
+/// Parameters to configure the pruning of the event log.
+#[derive(Debug, Copy, Clone)]
+pub struct Params {
+    /// Soft lock on the maximum number of events the event log can hold.
+    ///
+    /// If the number of events in the log exceeds this value, the log
+    /// will be pruned.
+    pub max_log_events: usize,
+    /// Soft lock on the number of entries the event log can hold.
+    ///
+    /// If the difference between the newest log entry and the oldest's
+    /// block heights is greater than this value, the log will be pruned.
+    pub log_block_height_diff: u64,
+}
+
+impl Default for Params {
+    fn default() -> Self {
+        // TODO: tune the default params
+        Self {
+            max_log_events: 50000,
+            log_block_height_diff: 1000,
+        }
+    }
+}
+
+/// Instantiates a new event log and its associated machinery.
+///
+/// General usage flow:
+///
+///   1. Spawn a new asynchronous task, with a [`Logger`]
+///      running on a loop.
+///   2. Send new events to the [`Logger`] with a [`LogEntrySender`].
+///      This will alter the state of the [`EventLog`].
+///   3. Concurrently, other asynchronous tasks may access the
+///      [`EventLog`] to check for new events.
+pub fn new(params: Params) -> (EventLog, Logger, LogEntrySender) {
+    let (tx, rx) = mpsc::unbounded_channel();
+
+    let log = EventLog::new(params);
+    let logger = Logger {
+        receiver: rx,
+        log: log.clone(),
+    };
+    let sender = LogEntrySender { sender: tx };
+
+    (log, logger, sender)
+}
+
+/// Represents an entry in the event log.
+#[derive(Debug, Clone)]
+pub struct LogEntry {
+    /// The block height at which we emitted the events.
+    pub block_height: BlockHeight,
+    /// The events emitted by a `FinalizeBlock` call.
+    pub events: Vec<Event>,
+}
+
+/// Represents a node in the linked list of log entries.
+#[derive(Debug)]
+struct LogNode {
```

Comment:
Linked lists are terrible data structures :sweat_smile: I think we can just use a vec here and then all the iter definitions simplifiy. The only question is the snapshot, which I haven't figured the purpose of yet.

Reply:
Or better yet some sort of automatically sorted data structure, like a btreeset.

Reply:
closing this after our discussion on a call